1 6 2 spec changes

PAldore opened the pull request:
Made changes to accommodate 1.6.2 specification uplift

This includes adding/updating existing steps, feature and data files with the addition of new patients

PAldore committed:
PA: Added patient 37,38 and 39 to NHSNOMAP
PA: Added patient 37,38 and 39 to Test Suite Pre requisite
PA: Updated StructuredAllergySteps for 1.6.2 change for resolved allergies to be returned as transfer-degraded drug allergies
PA:  Updated StructuredInvestigationSteps for 1.6.2 Update to Specimen collection.extension[fastingStatus] to include coded concept
PA:  Added steps to StructuredMedicationSteps for 1.6.2 Update for dose syntax to support Information Standards Notice DAPB401 for legacy medications
PA:  Added scenario to StructuredMedication Feature for 1.6.2 Update for dose syntax to support Information Standards Notice DAPB401 for legacy medications
PA:  Added steps to StructuredMedicationSteps for 1.6.2 Update for dose syntax to support Information Standards Notice DAPB401 for structured medications
PA:  Added scenario to StructuredMedication feature for 1.6.2 Update for dose syntax to support Information Standards Notice DAPB401 for structured medications
PA:  Added steps to StructuredMedicationSteps for 1.6.2 Update to MedicationRequest.medication and MedicationStatement.medication datatype
PA:  Added scenario to StructuredMedicationSteps for 1.6.2 Update to MedicationRequest.medication and MedicationStatement.medication datatype
PA:  Added steps to StructuredMultipleSteps for 1.6.2 Update to Resources may contain a ‘no disclosure to patient’ security label when "retrieving" or "migrating" a patients record
PA:  Added scenario to StructuredMigrate for 1.6.2 Update to Resources may contain a ‘no disclosure to patient’ security label when "retrieving" or "migrating" a patients record
PA:  Added scenario to StructuredMultipleRequests for 1.6.2 Update to Resources may contain a ‘no disclosure to patient’ security label when "retrieving" or "migrating" a patients record
PA: Updated patient in StructuredAllergies feature
PA: Updated patient in StructuredConsultations feature
PA: Updated patient in StructuredInvestigations feature
PA: Removed scenarios in StructuredProblems feature for 1.6.2 change for Remove problem significance filter
PA: Added scenario in StructuredProblems feature for 1.6.2 change for Ensure OperationDefinition aligns with profile

PAldore committed:
PA: Added comment and updated patient in StructuredMigrate feature for change where applicable resources may contain a ‘no disclosure to patient’ security label

PA: Updated comment  in StructuredMultiple feature for change where applicable resources may contain a ‘no disclosure to patient’ security label

PAldore committed:
PA: Reworded test scenarios to match expected

st3v3nhunt commented on GPConnect.Provider.AcceptanceTests/Steps/StructuredAllergySteps.cs:
```diff
@@ -149,6 +149,31 @@ public void TheListOfAllergyIntolerancesShouldBeValid()
+
+                                        });
+
+									if (allergyIntolerance.Code.Text != null)
```
In theory there shouldn't be any cases where the text is null as it should be coming from the original coded concept's description.

PAldore replied:
Removed text is null and passed through original coded concept's description for validation

st3v3nhunt commented on GPConnect.Provider.AcceptanceTests/Steps/StructuredAllergySteps.cs:
```diff
@@ -149,6 +149,31 @@ public void TheListOfAllergyIntolerancesShouldBeValid()
+
+									if (allergyIntolerance.Code.Text != null)
+									{
+										allergyIntolerance.Code.Text.ToString().ShouldStartWith("`Resolved");
```
Could this check be extended to use `Regex.IsMatch` with a pattern of e.g. `Resolved '.+' original code .+\.`?

PAldore replied:
Updated existing validation to use Regex.IsMatch

st3v3nhunt commented on GPConnect.Provider.AcceptanceTests/Steps/StructuredInvestigationsSteps.cs:
```diff
@@ -347,6 +347,36 @@ public void ThenIChecktheSpecimensareValid()
+                {
+                    if (specimen.Collection.Extension != null)
+                    {
+                        specimen.Collection.Extension.ForEach(extension =>
```
I know the extension could have many elements but I think reality it'll only have one so you could just check if there is one rather than for each.

st3v3nhunt commented on GPConnect.Provider.AcceptanceTests/Steps/StructuredInvestigationsSteps.cs:
```diff
@@ -347,6 +347,36 @@ public void ThenIChecktheSpecimensareValid()
+                        specimen.Collection.Extension.ForEach(extension =>
+                        {
+                            //Check extension.url
+                            extension.Url.ShouldNotBeNullOrEmpty("Url Is Null or Empty - Should be populated");
```
The URL has a fixed value of `https://fhir.hl7.org.uk/STU3/StructureDefinition/Extension-CareConnect-FastingStatus-1` so this could validated.

PAldore replied:
Added validation for fixed value https://fhir.hl7.org.uk/STU3/StructureDefinition/Extension-CareConnect-FastingStatus-1

st3v3nhunt commented on GPConnect.Provider.AcceptanceTests/Steps/StructuredInvestigationsSteps.cs:
```diff
@@ -347,6 +347,36 @@ public void ThenIChecktheSpecimensareValid()
+                            coding.Count.ShouldBeGreaterThan(0, "coding Is Null or Empty - Should be populated");
+                            coding.ForEach(_coding =>
+                            {
+                                _coding.System.ShouldNotBeNullOrEmpty();
```
Each of the elements of the codeable concept could be tested specifically, check the example in the spec for the values - 
https://simplifier.net/guide/gpconnect-data-model/Home/FHIR-Assets/All-assets/Profiles/Profile--CareConnect-GPC-Specimen-1?version=current#collection.extension-fastingStatus

PAldore replied:
Added validation to test values

st3v3nhunt commented on GPConnect.Provider.AcceptanceTests/Steps/StructuredInvestigationsSteps.cs:
```diff
@@ -347,6 +347,36 @@ public void ThenIChecktheSpecimensareValid()
+                            });
+
+                            //Check value
+                            extension.Value.ShouldNotBeNull("Value Is Null or Empty - Should be populated");
```
Is this checking the value of codeable concept's `text` property? If so, it could be updated to test for the text being `Patient was fasting prior to the procedure.`.

PAldore replied:
Added switch to validate text

st3v3nhunt commented on GPConnect.Provider.AcceptanceTests/Steps/StructuredMultipleSteps.cs:
```diff
@@ -339,5 +343,98 @@ public void checkTheResponseDoesNotContainAnOperationalOutcome()
+							coding.Code.ShouldNotBeNull();
+							coding.Code.ShouldBe("NOPAT");
+							coding.Display.ShouldNotBeNull();
+							coding.Display.ToLower().ShouldBe("no disclosure to patient, family or caregivers without attending provider's authorization");
```
Might want to consider extracting this logic into a reusable function given it is repeated a couple times within this method.

PAldore replied:
Created function to deal with repetitive steps

st3v3nhunt commented on GPConnect.Provider.AcceptanceTests/StructuredProblems.feature:
Rather than leaving the tests in and commented out, you might want to consider deleting them as the history is available within the repo.

PAldore replied:
Deleted tests

st3v3nhunt reviewed (changes requested):
Just a few suggestions and queries.

PAldore committed:
PA: Added FastingStatus ext to FhirConst
PA: Changed StructuredAllergySteps validation to Regex.IsMatch from StartsWith
PA: Updated StructuredInvestigationStep Validate Specimen from ShouldNotBeNullOrEmpty and added a switch statement to validate values returned
PA: Added function to StrucutredMultipleSteps to validate security label instead of repetetive validation, passed though to Step,removed List foreach and updated count

PAldore committed:
PA: Changed StructuredMultipleSteps for no disclosure to patient’ security label validation to check count

PAldore committed:
PA: Removed test  scenarios due to removal of of problem significance filter 1.6.2

st3v3nhunt reviewed (approved):
LGTM 👍

PAldore committed:
PA: Added new feature file for 1.6.2 spec changes
PA: Added existing scenarios into new feature file for execution against 1.6.2 changes
PA: Updated scenario name to correctly reflect medication being returned
